codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java

```diff
@@ -480,6 +487,66 @@ private static boolean contains(Http2Headers headers, CharSequence name) {
         return false;
     }
 
+    private static void verifyValidHeaderValue(int streamId, CharSequence name, CharSequence value)
+            throws Http2Exception {
+        // Validate value to field-content rule.
+        //  field-content  = field-vchar [ 1*( SP / HTAB ) field-vchar ]
+        //  field-vchar    = VCHAR / obs-text
+        //  VCHAR          = %x21-7E ; visible (printing) characters
+        //  obs-text       = %x80-FF
+        //  SP             = %x20
+        //  HTAB           = %x09 ; horizontal tab
+        //  See: https://datatracker.ietf.org/doc/html/rfc7230#section-3.2
+        //  And: https://datatracker.ietf.org/doc/html/rfc5234#appendix-B.1
+        int length = value.length();
+        if (length == 0) {
+            return;
+        }
+        if (value instanceof AsciiString) {
+            verifyValidHeaderValueAsciiString(streamId, name, (AsciiString) value);
+        } else {
+            verifyValidHeaderValueCharSequence(streamId, name, value);
+        }
+    }
+
+    private static void verifyValidHeaderValueAsciiString(int streamId, CharSequence name, AsciiString value)
+            throws Http2Exception {
+        final byte[] array = value.array();
+        final int start = value.arrayOffset();
+        int b = array[start] & 0xFF;
+        if (b < 0x21 || b == 0x7F) {
+            throw illegalHeaderValue(streamId, name, b, 0);
+        }
+        int length = value.length();
+        for (int i = start + 1; i < length; i++) {
+            b = array[i] & 0xFF;
+            if (b < 0x20 && b != 0x09 || b == 0x7F) {
+                throw illegalHeaderValue(streamId, name, b, i - start);
+            }
+        }
+    }
+
+    private static void verifyValidHeaderValueCharSequence(int streamId, CharSequence name, CharSequence value)
```

Comment:
This method isn't tested. For all test cases you created, consider running them for both `AsciiString` and a regular `String`.

Reply:
As we are in the decoder private APIs, and the decoder only produce AsciiString, we can alternatively pin that down in the API and just not generalize across all CharSequences.